MERL-617: plugin guide (#1397)
wpengine/faustjs: internal/faustjs.org/docs/guides/plugins.mdx

---
slug: /guides/plugins
title: Plugins
description: Learn about Faust Plugins
---

This guide covers how to set up Faust plugins and how to use them in your site.

## Experimental Config Object

> The experimental config object is used to setup plugins via the `experimentalPlugins` property. It's a property that takes an array of plugin class instances. You or the end user can then implement the plugin by adding it to the `experimentalPlugins` property in the App's `faust.config.js` file. See the example below:

Reviewer: ```suggestion
The experimental config object is used to set up plugins via the `experimentalPlugins` property. It's a property that takes an array of plugin class instances. You or the end user can then implement the plugin by adding it to the `experimentalPlugins` property in the App's `faust.config.js` file. See the example below:
```

```js title="faust.config.js"
import { setConfig } from '@faustwp/core';
import templates from './wp-templates';
import possibleTypes from './possibleTypes.json';

> import { MyPlugin } from './MyPlugin.tsx';

Reviewer 2: You cannot import a `tsx` file in javascript so this would fail.

/**
 * @type {import('@faustwp/core').FaustConfig}
 **/
export default setConfig({
  templates,
  experimentalPlugins: [new MyPlugin()],
  possibleTypes,
});
```

## Plugin Options

:::note
The Faust plugins utilize the `@wordpress/hooks` library found [here](https://www.npmjs.com/package/@wordpress/hooks).
:::

When you create a plugin you will use the `@wordpress/hooks` library as well as [Faust filters](/docs/plugin-system/filters).

## Create a plugin

To get started, follow our guide on [Creating a plugin](/docs/plugin-system/creating-a-plugin)

## Example Usage

### Tapping into Apollo

[Apollo Client Options](https://www.apollographql.com/docs/react/api/core/ApolloClient/)

Below is an example using `addFilter` that modifies the `ApolloClientOptions`.

> ```js

Reviewer 2: ```suggestion
```ts
```

import {
  ApolloClientOptions,
  NormalizedCacheObject,
} from '@apollo/client';

addFilter(
    hookName: 'apolloClientOptions',
    namespace: string,
    callback: (
      apolloClientOptions: ApolloClientOptions<NormalizedCacheObject>,
      context: Record<string, never>,
    ) => ApolloClientOptions<NormalizedCacheObject>,
    priority?: number | undefined,
  ): void;
```

> [InMemoryCache](https://www.apollographql.com/docs/react/api/cache/InMemoryCache/)

Reviewer: I would suggest adding clarity on this link:

```suggestion
`InMemoryCache` is the cache used by almost every instance of ApolloClient. Read more about `InMemoryCache` [here.](https://www.apollographql.com/docs/react/api/cache/InMemoryCache/)
```

Below is an example using `addFilter` that modifies the `InMemoryCache`.

> ```js

Reviewer 2: ```suggestion
```ts
```

import {
  InMemoryCacheConfig,
} from '@apollo/client';

addFilter(
    hookName: 'apolloClientInMemoryCacheOptions',
    namespace: string,
    callback: (
      inMemoryCacheObject: InMemoryCacheConfig,
      context: Record<string, never>,
    ) => InMemoryCacheConfig,
    priority?: number | undefined,
  ): void;
```

### Persisted Queries

Here we show an example of a persisted query using a Faust plugin. You can find this example on GitHub [here](https://github.com/wp-graphql/wpgraphql.com/blob/master/src/plugins/PersistedQueriesPlugin.js).

> It uses the `addFilter` inside a Faust plugin called `PersistedQueriesPlugin`. It filters `apolloClientOptions` and pulls the link as `existingLink`. If the `existingLink` is an instance of `HttpLink` it adds the `existingLink` to the `persistedQueriesLink`. If it's not an instance of `HttpLink` it adds the `httpLink` to the `persistedQueriesLink`.

Reviewer: ```suggestion
It uses the `addFilter` inside a Faust plugin called `PersistedQueriesPlugin`. It filters `apolloClientOptions` and pulls the link as `existingLink`. If the `existingLink` is an instance of `HttpLink`, it adds the `existingLink` to the `persistedQueriesLink`. If it's not an instance of `HttpLink`, it adds the `httpLink` to the `persistedQueriesLink`.
```

```js title="PersistedQueriesPlugin.js"
import { createPersistedQueryLink } from '@apollo/client/link/persisted-queries';
import { HttpLink } from '@apollo/client';
import { sha256 } from 'crypto-hash';
import { getGraphqlEndpoint } from '@faustwp/core/dist/mjs/lib/getGraphqlEndpoint';

const httpLink = new HttpLink({ uri: getGraphqlEndpoint() });
const persistedQueriesLink = createPersistedQueryLink({
  sha256,
  useGETForHashedQueries: true,
});

class PersistedQueriesPlugin {
  apply({ addFilter }) {
    addFilter('apolloClientOptions', 'faust', (apolloClientOptions) => {
      const existingLink = apolloClientOptions?.link;
      return {
        ...apolloClientOptions,
        link:
          existingLink instanceof HttpLink
            ? persistedQueriesLink.concat(existingLink)
            : persistedQueriesLink.concat(httpLink),
      };
    });
  }
}

export default PersistedQueriesPlugin;
```

## Additional Filters

> For further learning about the filters check out the [Faust Filters](https://faustjs.org/docs/plugin-system/filters).

Reviewer: ```suggestion
For further learning about Faust filters, check out the Faust Filters, [here.](https://faustjs.org/docs/plugin-system/filters).
```

---
slug: /guides/plugins
title: Plugins
description: Learn about Faust Plugins
---

This guide covers how to set up Faust plugins and how to use them in your site.

> Prerequisites

Reviewer: ```suggestion
Prerequisites:
```

```js title="PersistedQueriesPlugin.js"
import { createPersistedQueryLink } from '@apollo/client/link/persisted-queries';
import { HttpLink } from '@apollo/client';
import { sha256 } from 'crypto-hash';
import { getGraphqlEndpoint } from '@faustwp/core/dist/mjs/lib/getGraphqlEndpoint';

const httpLink = new HttpLink({ uri: getGraphqlEndpoint() });
const persistedQueriesLink = createPersistedQueryLink({
  sha256,
  useGETForHashedQueries: true,
});

class PersistedQueriesPlugin {
  apply({ addFilter }) {
    addFilter('apolloClientOptions', 'faust', (apolloClientOptions) => {
      const existingLink = apolloClientOptions?.link;
      return {
        ...apolloClientOptions,
        link:
          existingLink instanceof HttpLink
            ? persistedQueriesLink.concat(existingLink)
            : persistedQueriesLink.concat(httpLink),
      };
    });
  }
}

export default PersistedQueriesPlugin;
```

## Additional Filters

> For further learning about Faust filters, check out the Faust Filters, [here.](https://faustjs.org/docs/plugin-system/filters).

Reviewer: ```suggestion
For further learning about Faust filters, check out the Faust Filters, [here](https://faustjs.org/docs/plugin-system/filters).
```
